### PR TITLE
Feature/authentication logic

### PR DESCRIPTION
**Pull request by DavMovsesyan2002**

### CODE REVIEW

**Review by rikarsen — CHANGES_REQUESTED** [review]

File: src/manager/axiosInstance.ts [hunk 1]
```diff
@@ -18,10 +18,12 @@ class RequestManager {
       const requestConfig = config
 
       // TODO: update this after login endpoint integration
-      const idToken = 'TOKENHERE'
+      const idToken = localStorage.getItem('accessToken')
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/axiosInstance.ts:21]: Yep

File: src/redux/users/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+const login = (params: ILoginListParams) => async (dispatch: AppDispatch) => {
+  try {
+    dispatch(setIsLoading(true))
+
+    const response = await API.authLogin.login(params)
+
+    localStorage.setItem('accessToken', response.data.data.accessToken)
+    dispatch(setError(false))
```
Comment (ArthurO2002) [inline src/redux/users/middleware.ts:26]: Error state should not be boolean

File: src/redux/users/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (response.status === 200) {
+      dispatch(setSuccessfully(true))
+    }
```
Comment (ArthurO2002) [inline src/redux/users/middleware.ts:30]: if request is still in try it's a happy flow, no need to check statusCode

File: src/redux/users/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  try {
+    dispatch(setIsLoading(true))
+
+    const response = await API.authLogin.login(params)
+
+    localStorage.setItem('accessToken', response.data.data.accessToken)
+    dispatch(setError(false))
+
+    if (response.status === 200) {
+      dispatch(setSuccessfully(true))
```
Comment (ArthurO2002) [inline src/redux/users/middleware.ts:29]: We need to store user data in redux, if it's there then request is successfull so no need of setSuccessfully

File: src/redux/users/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    dispatch(setIsLoading(true))
+
+    const response = await API.authRegister.register(params)
+
+    dispatch(setRedirectionState({ path: '/login', params: '', apply: true }))
+
+    dispatch(setError(false))
+
+    if (response.status === 204) {
+      dispatch(setSuccessfully(true))
```
Comment (ArthurO2002) [inline src/redux/users/middleware.ts:50]: Same here

File: src/utils/isAuthenticated.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export const isAuthenticated = () => {
+  const token = localStorage.getItem('accessToken')
+
```
Comment (ArthurO2002) [inline src/utils/isAuthenticated.ts:3]: Rename file to authHelpers

File: src/pages/login/index.tsx [hunk 7]
```diff
@@ -1,13 +1,30 @@
...
+  const [email, setEmail] = useState<string>('')
+  const [password, setPassword] = useState<string>('')
+
+  const handleLoginFunc = () => {
+    dispatch(usersMiddleware.login({ email, password }))
+  }
+
+  useEffect(() => {
+    if (successfully) {
+      push('/profile')
```
Comment (ArthurO2002) [inline src/pages/login/index.tsx:25]: redirect should be in middleware in success case

**Review by ArthurO2002 — CHANGES_REQUESTED** [review]

File: src/manager/API.ts [hunk 8]
```diff
@@ -1,7 +1,11 @@
+import loginManager from '@axios/authentication/login/loginManager'
+import registerManager from '@axios/authentication/register/registerManager'
 import exampleManager from '@axios/example/exampleManager'
 
 const API = {
   example: exampleManager,
+  authLogin: loginManager,
+  authRegister: registerManager,
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/API.ts]: Fixed

File: src/manager/authentication/login/loginHelpers.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  id: string
+  page: number
+  value: string
+}
+
+const loginFunction = (params: ExampleListParams) => {
+  const { page, value } = params
+
+  return value[page]
+}
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/authentication/login/loginHelpers.ts]: Fixed

File: src/manager/authentication/login/loginHelpers.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+const loginHelpers = {
+  loginFunction,
+}
+
+export default loginHelpers
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/authentication/login/loginHelpers.ts]: Fixed

File: src/manager/authentication/register/registerHelpers.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+const registerFunction = (params: ExampleListParams) => {
+  const { page, value } = params
+
+  return value[page]
+}
+
+const registerHelpers = {
+  registerFunction,
+}
+
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/authentication/register/registerHelpers.ts]: Fixed

File: src/manager/authentication/register/registerManager.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+const baseURL = '/users'
+
+const axiosInstance = Axios()
+const registerManager = {
+  axiosInstance,
+  register(params: IRegisterListParams) {
+    return axiosInstance.post<IRegisterResponse, IAxiosResponse>(
+      `${baseURL}/v1/auth/register`,
+      params
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/authentication/register/registerManager.ts]: Fixed

File: src/manager/axiosTypes.ts [hunk 13]
```diff
@@ -1,10 +1,10 @@
 import { AxiosResponse } from 'axios'
 
-export interface IAxiosResponse<T> extends AxiosResponse {
-  data: IData<T>
+export interface IAxiosResponse extends AxiosResponse {
+  data: AxiosResponse
 }
-export interface IAxiosResponsePaginated<T> extends AxiosResponse {
-  data: IPaginatedData<T>
+export interface IAxiosResponsePaginated extends AxiosResponse {
```
[the remark this answers is not in the document]
Comment (ArthurO2002) [reply src/manager/axiosTypes.ts]: Fixed

File: src/redux/users/middleware.ts (hunk 2 above)
Comment (ArthurO2002) [reply src/redux/users/middleware.ts]: Fixed

File: src/redux/users/middleware.ts (hunk 3 above)
Comment (ArthurO2002) [reply src/redux/users/middleware.ts]: Fixed

File: src/redux/users/middleware.ts (hunk 4 above)
Comment (ArthurO2002) [reply src/redux/users/middleware.ts]: Fixed

File: src/redux/users/middleware.ts (hunk 5 above)
Comment (ArthurO2002) [reply src/redux/users/middleware.ts]: Fixed

File: src/utils/isAuthenticated.ts (hunk 6 above)
Comment (ArthurO2002) [reply src/utils/isAuthenticated.ts:3]: Fixed

File: src/pages/login/index.tsx (hunk 7 above)
Comment (ArthurO2002) [reply src/pages/login/index.tsx]: Fixed

**Review by ArthurO2002 — APPROVED** [review]

**Review by ArthurO2002 — APPROVED** [review]

**Review by ArthurO2002 — DISMISSED** [review]

File: src/manager/authentication/authManagerTypes.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  email: string
+  password: string
+  remember: boolean
+}
+
+// TODO: change response interface to match your backend
+export interface ISignUpResponse {
+  accessToken: string
+  fullName: string
+}
```
Comment (ArthurO2002) [inline src/manager/authentication/authManagerTypes.ts:17]: Look into network request and see what BE is providing

File: src/pages/register/index.tsx [hunk 15]
```diff
@@ -33,31 +30,46 @@ export const Register = () => {
...
-        <input
-          type="text"
-          id="name"
-          placeholder="Full Name"
-          className="mb-5 w-full rounded border border-gray-light px-5 py-2.5 outline-0 placeholder:text-base placeholder:text-black"
-        />
         <div className="mb-5 w-full">
           <input
+            value={fullName}
+            onChange={(e) => setFullName(e.target.value)}
```
Comment (ArthurO2002) [inline src/pages/register/index.tsx:36]: move to function

File: src/pages/register/index.tsx [hunk 16]
```diff
@@ -1,30 +1,27 @@
...
 export const Register = () => {
+  const { isSignUpLoading, error } = useAppSelector(usersSelector.user)
   const [show, setShow] = useState(false)
-  const error = false
+  const [fullName, setFullName] = useState<string>('')
+  const [email, setEmail] = useState<string>('')
+  const [password, setPassword] = useState<string>('')
+  const [confirmPassword, setConfirmPassword] = useState<string>('')
+
+  const handleRegisterFunc = () => {
```
Comment (ArthurO2002) [inline src/pages/register/index.tsx:18]:
```suggestion
  const handleRegister = () => {
```

File: src/pages/register/index.tsx [hunk 17]
```diff
@@ -109,7 +125,16 @@ export const Register = () => {
             Remember me
           </label>
         </div>
-        <Button size="fl">Next</Button>
+        <Button
+          disabled={isSignUpLoading}
+          size="fl"
+          onClick={() => handleRegisterFunc()}
```
Comment (ArthurO2002) [inline src/pages/register/index.tsx:131]:
```suggestion
          onClick={handleRegisterFunc}
```

File: src/pages/register/index.tsx [hunk 18]
```diff
@@ -109,7 +125,16 @@ export const Register = () => {
...
+        >
+          Create Account
+        </Button>
+        <div className="w-full">
+          {error ? <p className="mt-2.5 text-red">Something went wrong</p> : null}
```
Comment (ArthurO2002) [inline src/pages/register/index.tsx:136]: We need to show error from BE if it's there

File: src/redux/slices/users/middleware.ts [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    dispatch(setIsAuthenticated(true))
+    dispatch(setError(null))
+  } catch (error) {
+    dispatch(setError(error as Error))
+  } finally {
+    dispatch(setSignInLoading(false))
+  }
+}
+
+const IsAuthenticated = () => async (dispatch: AppDispatch) => {
```
Comment (ArthurO2002) [inline src/redux/slices/users/middleware.ts:36]:
```suggestion
const isAuthenticated = () => async (dispatch: AppDispatch) => {
```

**Review by ArthurO2002 — CHANGES_REQUESTED** [review]

File: src/pages/register/index.tsx (hunk 17 above)
Comment (DavMovsesyan2002) [reply src/pages/register/index.tsx:131]: Done

File: src/pages/register/index.tsx (hunk 16 above)
Comment (DavMovsesyan2002) [reply src/pages/register/index.tsx:18]: Done

File: src/pages/login/index.tsx [hunk 20]
```diff
@@ -80,7 +83,20 @@ export const Login = () => {
             Remember me
           </label>
         </div>
-        <Button size="fl">Next</Button>
+        <Button
+          className="mb-7"
+          size="fl"
+          disabled={isSignInLoading}
+          onClick={() => handleLoginFunc()}
```
[the remark this answers is not in the document]
Comment (DavMovsesyan2002) [reply src/pages/login/index.tsx:90]: Done

File: src/pages/register/index.tsx (hunk 18 above)
Comment (DavMovsesyan2002) [reply src/pages/register/index.tsx:136]: Done

**Review by ArthurO2002 — APPROVED** [review]

**Review by ArthurO2002 — DISMISSED** [review]

File: src/manager/authentication/authManagerTypes.ts (hunk 14 above)
Comment (DavMovsesyan2002) [reply src/manager/authentication/authManagerTypes.ts]: Done

**Review by rikarsen — APPROVED** [review]

**Review by rikarsen — APPROVED** [review]

**Pull request merged**